### PR TITLE
Experimental fix for serious OpenVPN flaw

### PR DESCRIPTION
**Pull request by holta**
This is _a_ possible fix to a structural flaw that's existed in our OpenVPN code for years, and manifested itself as OpenVPN accidentally turning on, even when @tim-moody & @kananigit had set their [/etc/iiab/local_vars.yml](http://wiki.laptop.org/go/IIAB/local_vars.yml) to: 
```
openvpn_install: True
openvpn_enabled: False
```
Testing ongoing...

### CODE REVIEW

**Comment by holta** [comment]
@kananigit will test this later this evening with a fresh install of IIAB 6.6/master (http://download.iiab.io/6.6) and report back.

It might not be the final word.

**Comment by holta** [comment]
The architectural flaw in our 5-year old design appears to be:

CHILD service "openvpn@xscenet" was being stopped but having having no long-term effect in many conditions...and then after reboot PARENT service "openvpn" was auto-starting it (CHILD service "openvpn@xscenet" and actual tunnel) regardless what Ansible had done prior to reboot.

As a result of these files:
- **/etc/default/openvpn** (implies AUTOSTART="all")
- **/etc/init.d/openvpn** (has AUTOSTART="all")
- **/etc/openvpn/xscenet.conf** (our VPN connection)
- **/etc/network/if-up.d/openvpn** (appears to auto-start xscenet.conf)
- **/lib/systemd/systemd-sysv-install** (sets /etc/rc*.d/S|K01openvpn e.g. when "systemctl enable openvpn")

**Comment by holta** [comment]
Tested on a fresh install of a (MIN-sized) Debian 9.5 VM.

ASIDE: subsequently OpenVPN appears as it should when you run:
```
cd /opt/iiab/iiab
./install-support
```

**Comment by holta** [comment]
@kananigit reconfirms this works (i.e. OpenVPN connection does not appear) after he installed a MEDIUM-sized on RPi3.

**Comment by georgejhunt** [comment]
I don't know how I can help you.  I do not know anything about parent, and
child openvpn instances. When I google, looking for assistance, the
following seems to be authoritative:

https://github.com/OpenVPN/openvpn/tree/master/distro/systemd

I find mentions of separate instances of openvpn servers specified by
config file name:

 - server configurations:
    # systemctl $OPER openvpn-server@$CONFIGNAME


But I do not find any mention of parent/child. I do not know what a raw
openvpn -- see line 168 in roles/openvpn/tasks/main.yml (as opposed to
openvpn@xscenet.service) is referring to.

**Comment by holta** [comment]
```
root@box:/lib/systemd/system# ls -l openvpn*
-rw-r--r-- 1 root root 707 Jul 18  2017 openvpn-client@.service
-rw-r--r-- 1 root root 780 Jul 18  2017 openvpn-server@.service
-rw-r--r-- 1 root root 320 Jul 18  2017 openvpn.service          ("PARENT service")
-rw-r--r-- 1 root root 894 Jul 18  2017 openvpn@.service          ("CHILD service")
```
Example PARENT service:
```
root@box:/opt/iiab/iiab# systemctl status openvpn
● openvpn.service - OpenVPN service
   Loaded: loaded (/lib/systemd/system/openvpn.service; enabled; vendor preset: enabled)
   Active: active (exited) since Sun 2018-09-02 22:40:42 EDT; 1h 52min ago
  Process: 612 ExecStart=/bin/true (code=exited, status=0/SUCCESS)
 Main PID: 612 (code=exited, status=0/SUCCESS)
    Tasks: 0 (limit: 4915)
   CGroup: /system.slice/openvpn.service

Sep 02 22:40:42 box.lan systemd[1]: Starting OpenVPN service...
Sep 02 22:40:42 box.lan systemd[1]: Started OpenVPN service.
```
Example CHILD service, auto-launched by PARENT service above:
```
root@box:/opt/iiab/iiab# systemctl status openvpn@xscenet
● openvpn@xscenet.service - OpenVPN connection to xscenet
   Loaded: loaded (/lib/systemd/system/openvpn@.service; disabled; vendor preset: enabled)
   Active: active (running) since Sun 2018-09-02 22:40:38 EDT; 1h 52min ago
     Docs: man:openvpn(8)
           https://community.openvpn.net/openvpn/wiki/Openvpn23ManPage
           https://community.openvpn.net/openvpn/wiki/HOWTO
  Process: 350 ExecStart=/usr/sbin/openvpn --daemon ovpn-xscenet --status /run/openvpn/xscenet.status 10 --cd /etc/openvpn --config /etc/openvpn/xscenet.conf --writepid
 Main PID: 372 (openvpn)
    Tasks: 2 (limit: 4915)
   CGroup: /system.slice/system-openvpn.slice/openvpn@xscenet.service
           ├─ 372 /usr/sbin/openvpn --daemon ovpn-xscenet --status /run/openvpn/xscenet.status 10 --cd /etc/openvpn --config /etc/openvpn/xscenet.conf --writepid /run/o
           └─1174 /usr/bin/ncat -l -k -p1705 --exec /bin/echo HANDLE = deb9.5-MIN-2018-09-02|UUID = 01e10c04-a13e-4756-bca9-fde9a6edcc55

Sep 02 22:40:36 box.lan systemd[1]: Starting OpenVPN connection to xscenet...
Sep 02 22:40:38 box.lan systemd[1]: openvpn@xscenet.service: PID file /run/openvpn/xscenet.pid not readable (yet?) after start: No such file or directory
Sep 02 22:40:38 box.lan systemd[1]: Started OpenVPN connection to xscenet.
```

**Comment by holta** [comment]
Root cause of the flaw: (tunnels were being spawned erroneously)

"apt install openvpn" (i.e. [lines 1-6 of roles/openvpn/tasks/main.yml](https://github.com/iiab/iiab/blob/master/roles/openvpn/tasks/main.yml#L1-L6)) automatically enables OpenVPN the very moment it's installed (thereby creating CHILD services like "openvpn@xscenet" for any tunnel.conf file like xscenet.conf that exists in /etc/openvpn during the next boot...)

_...against the desires of the implementer/operator who was clearly trying to avoid that with `openvpn_enabled: False` in local_vars.yml !_

**Comment by holta** [comment]
CLARIF: there are of course other ways this gotcha could be fixed.

But disabling the PARENT service openvpn.service (as done by PR #1079 here) provides a simple/foolproof way to verify OpenVPN tunnels are "never again" accidentally turned on.

**Comment by holta** [comment]
@jvonau further clarifies:

> [it looks like /etc/openvpn/xscenet.conf] is being started by /etc/network/if-up.d/openvpn

> [Derived from /etc/default/openvpn] /etc/init.d/openvpn has AUTOSTART="all" and systemd is calling  /lib/systemd/systemd-sysv-install when enabling the openvpn service so with any *.conf file the [daemon] fires off with that config file...

>> /etc/network/if-up.d/openvpn calls xscenet.conf regardless of NM right?

> I'm not sure, think /etc/init.d/openvpn is handling the process as systemd's openvpn.service file does nothing... 

> but either one could start the tunnel but if-up is suppose to be [deprecated]

**Comment by jvonau** [comment]
This is very much like dhcpd where during the install we disable the stock systemd services and enable our own version via a systemd service file. Think what should be done is to unconditionally disable openvpn.service at the systemd level to disable the sysinit /etc/init.d/openvpn component that systemd calls when en/disabling openvpn(alters /etc/rc*.d/K|S01openvpn) at first install and to use openvpn@xscenet.service exclusivity as was done in the past.

**Comment by tim-moody** [comment]
if this problem has existed (theoretically) for 5 years, why did no one ever see it before and 2 people saw it only on current master?  what has changed, the behavior of openvpn? or some other code? is openvpn a new version?

> Root cause of the flaw: (tunnels were being spawned erroneously)<

this is a symptom, not a root cause

can anyone verify that the symptom is present without this patch and absent with it?

**Comment by tim-moody** [comment]
> But I agree with Jerry. Let's disable the parent, and just manipulate the
child (openvpn@xscenet.service).

this PR seems headed in the opposite directrion.

any chance there is a spurious install-support somewhere in a script?

**Comment by holta** [comment]
> any chance there is a spurious install-support somewhere in a script?

When I spent a couple days removing bugs from [roles/openvpn/tasks/main.yml](https://github.com/iiab/iiab/blob/master/roles/openvpn/tasks/main.yml) and [roles/openvpn/templates/*](https://github.com/iiab/iiab/tree/master/roles/openvpn/templates) in August (documented in comments, to ease long-term maintenance) I introduced [one](https://github.com/iiab/iiab/commit/04be152c0d3d81306c2d5bdc3554f379728e6396#diff-b00b5723973858b765173eec0119c870R120) along the way, as I did not yet realize that parent service (openvpn) auto-starts child service (openvpn@xscenet) and so I made the mistake of assuming that bug was around for years (the others were, not this one).

Anyway, I now understand my mistake, and fixed that bug yesterday (in this PR #1079).

I'm not sure we really want the parent service unconditionally disabled when "openvpn_enabled: True" says exactly the opposite (in plain English!)  e.g. if a sysadmin wants to add other VPN connections (or anybody else who uses OpenVPN defaults in their career etc) I'd prefer that OpenVPN be enabled per its well-known defaults, rather than us customizing it in ways that will confuse others.

That's my hunch anyway, that we need to use industry defaults wherever possible here.  Of course both approaches work identically for amateur implementers (who would correctly ask why we're debating angels on the head of a pin :-)

**Comment by holta** [comment]
Back Refs: #995 PRs #996 #997 #998 #1000 #1001 #1003 #1004

**Comment by holta** [comment]
> I'm glad you found a solution that works. Do you think the parent instance can take the place of our crontab restarter?

I'm assuming that both work well together, wherever certain flaky environments may indeed need our crontab restarter.

i.e. I'm assuming crontab's hourly "systemctl start openvpn@xscenet" should work very well together with the higher-level "systemctl enable openvpn" at boot/etc.

Reason: I've quite aggressively tested "systemctl [start/restart/stop/disable/enable] openvpn@xscenet" and while useful in a pinch for tinkerers, these commands have only a transient effect, i.e. they have no long-term affect on the state during the next reboot or "systemctl restart openvpn".

PS I'm assuming it's not necessary to do "systemctl start openvpn" every hour, but there are many other things we could place in this hourly cron job as nec.

PPS I understand @jvonau favors a different approach (PR #1087) which feels very non-standard (to me) as if trying to customize a globally very standard tool in a non-standard way &mdash; to the point of alienating sysadmins who are already very familiar with OpenVPN's existing defaults & norms.  But let's talk more during Thursday's call (http://minutes.iiab.io)

**Comment by jvonau** [comment]
With peppering some extra logging on deb9.5-BIG-with-mods-2018-09-04 which shows that the vpn is being started when /etc/network/if-up.d/openvpn is called by networking service. I don't think a 'systemctl disable openvpn' is enough to disable the auto-launch of the vpn, but that depends on what /lib/systemd/systemd-sysv-install does when invoked by systemdctl, it might be writing disabled to /etc/defaults/openvpn but that needs to be confirmed.

oot@box:~# systemctl status networking
● networking.service - Raise network interfaces
   Loaded: loaded (/lib/systemd/system/networking.service; enabled; vendor preset: enabled)
   Active: active (exited) since Tue 2018-09-04 13:00:49 EDT; 1min 41s ago
     Docs: man:interfaces(5)
  Process: 394 ExecStart=/sbin/ifup -a --read-environment (code=exited, status=0/SUCCESS)
  Process: 392 ExecStartPre=/bin/sh -c [ "$CONFIGURE_INTERFACES" != "no" ] && [ -n "$(ifquery --read-en
 Main PID: 394 (code=exited, status=0/SUCCESS)
    Tasks: 0 (limit: 4915)
   CGroup: /system.slice/networking.service

Sep 04 13:00:42 box.lan systemd[1]: Starting Raise network interfaces...
Sep 04 13:00:46 box.lan ifup[394]: Loading netfilter rules...run-parts: executing /usr/share/netfilter-
Sep 04 13:00:46 box.lan ifup[394]: run-parts: executing /usr/share/netfilter-persistent/plugins.d/25-ip
Sep 04 13:00:46 box.lan ifup[394]: done.
Sep 04 13:00:46 box.lan ifup[394]: Loading netfilter rules...run-parts: executing /usr/share/netfilter-
Sep 04 13:00:46 box.lan ifup[394]: run-parts: executing /usr/share/netfilter-persistent/plugins.d/25-ip
Sep 04 13:00:46 box.lan ifup[394]: done.
Sep 04 13:00:48 box.lan ifup[394]: /etc/network/if-up.d/openvpn called...done.
Sep 04 13:00:49 box.lan ifup[394]: /etc/network/if-up.d/openvpn called...done.
Sep 04 13:00:49 box.lan systemd[1]: Started Raise network interfaces.

root@box:~# systemctl status openvpn@xscenet
● openvpn@xscenet.service - OpenVPN connection to xscenet
   Loaded: loaded (/lib/systemd/system/openvpn@.service; disabled; vendor preset: enabled)
   Active: active (running) since Tue 2018-09-04 13:00:43 EDT; 13min ago
     Docs: man:openvpn(8)
           https://community.openvpn.net/openvpn/wiki/Openvpn23ManPage
           https://community.openvpn.net/openvpn/wiki/HOWTO
  Process: 368 ExecStart=/usr/sbin/openvpn --daemon ovpn-xscenet --status /run/openvpn/xscenet.status 1
 Main PID: 406 (openvpn)
    Tasks: 2 (limit: 4915)
   CGroup: /system.slice/system-openvpn.slice/openvpn@xscenet.service
           ├─ 406 /usr/sbin/openvpn --daemon ovpn-xscenet --status /run/openvpn/xscenet.status 10 --cd 
           └─1105 /usr/bin/ncat -l -k -p1705 --exec /bin/echo HANDLE = deb9.5-BIG-with-mods-2018-09-04|

Sep 04 13:00:42 box.lan systemd[1]: Starting OpenVPN connection to xscenet...
Sep 04 13:00:43 box.lan systemd[1]: openvpn@xscenet.service: PID file /run/openvpn/xscenet.pid not read
Sep 04 13:00:43 box.lan systemd[1]: Started OpenVPN connection to xscenet.

root@box:~# systemctl status openvpn
● openvpn.service - OpenVPN service
   Loaded: loaded (/lib/systemd/system/openvpn.service; enabled; vendor preset: enabled)
   Active: active (exited) since Tue 2018-09-04 13:00:49 EDT; 3min 41s ago
  Process: 558 ExecStart=/bin/true (code=exited, status=0/SUCCESS)
 Main PID: 558 (code=exited, status=0/SUCCESS)
    Tasks: 0 (limit: 4915)
   CGroup: /system.slice/openvpn.service

Sep 04 13:00:49 box.lan systemd[1]: Starting OpenVPN service...
Sep 04 13:00:49 box.lan systemd[1]: Started OpenVPN service.
